### ISSUE TITLE
IPP Backend always sends IPP attributes when using file-parameter

### ISSUE THREAD

**Issue by SoerenBusse** [OpenPrinting/cups#671]
**Describe the bug**
When calling `/usr/lib/cups/backend/ipp` using the optional file parameter manually, the backend sends all options via IPP regardless of the `FINAL_CONTENT_TYPE` environment variable. When using `stdin` (like CUPS internal) the options are only sent when `FINAL_CONTENT_TYPE` is a PDF or an image (see: https://github.com/OpenPrinting/cups/blob/master/backend/ipp.c#L595)

**Why this is an issue**
We're using the print-solution PaperCut in our enviromnet using 3 printers with the following drivers:

| Printer | Driver  | Driver-Source | Protocol |
|---|---|---|---|
|  Samsung M5370LX  | Samsung M4370 5370 Series PS | ULD | IPP |
|  Samsung M4250LX | Samsung X4300 Series PS | ULD | IPP |
| HP Laserjet M553  | HP Color LaserJet M553 Postscript | HPLIP | IPP |

PaperCut registers itself as a backend in the device-uri and forward the job options and the spool file using the file-parameter to the IPP-Backend. The flow for PostScript (like in our case) looks like this (without knowing their code and by looking at the log files):
1. CUPS calls the PaperCut backend after filtering and passes the PostScript file via stdin
2. PaperCut writes the data from stdin into a spool file
3. PaperCut analyzes the Postscript file (e.g. to see if a document really contains color)
4. PaperCut calls the IPP Backend with a forced copy count of 1 using the file-parameter with the spool file

A print-request from a client with "Copy Count = 2", which is included as `NumCopies` in the generated PostScript file, looks like this, when the IPP-Backend is called using the File-Parameter.
```
%RBIBeginNonPPDFeature: *NumCopies 2
2/languagelevel where{pop languagelevel 2 ge}{false}ifelse
{1 dict begin/NumCopies exch def currentdict end setpagedevice}
{userdict/#copies 3 -1 roll put}ifelse
%RBIEndNonPPDFeature
```
![image](https://user-images.githubusercontent.com/7796567/233861018-93cbaa9f-0d95-4b16-aa41-f5a6eb00f614.png)

The IPP-Backend sets the appropriate IPP attributes from the passed options, except the `Copies` attribute, because it was forced to 1 by the PaperCut backend. In this case, the Samsung printer is only printing a single copy, although `NumCopies` is set to `2` inside the Postscript file. Looks like the Samsung IPP implementation defaults to `Copies = 1`, when there are **any** IPP attributes set and take precedence over the PostScript `NumCopies` value of 2.

The HP printer behaves differently with the same request. The HP printer prints `2` copies. 
| Printer | Copy parameter of IPP backend | Postscript NumCopies |Actually printed copies|
|---|---|---|---|
|  Samsung M5370LX  | 1 (priority) | 2 | 1 |
|  Samsung M4250LX | 1 (priority) | 2 | 1 |
| HP Laserjet M553  | 1 | 2 (priority) | 2 |

---

The behavior get more strange, when we disable the `ForceCopyCount = 1` in the PaperCut configuration, so the PaperCut backend forwards the correct copy count as parameter to the IPP backend. In this case, the HP printer is printing `n * n` 
 (IPP * Postscript) copies, while the Samsung printer now prints the correct amount of copies, because the `Copies` is correctly set to `2`.

| Printer | Copy parameter of IPP backend | Postscript NumCopies |Actually printed copies|
|---|---|---|---|
|  Samsung M5370LX  | 2 (priority) | 2 | 2 |
|  Samsung M4250LX | 2 (priority) | 2 | 2 |
| HP Laserjet M553  | 2 | 2 | 4 |

---

Without using the PaperCut backend, everything works correctly, because the CUPS filters are passing the generated Postscript as stdin to the IPP-Backend and the IPP-backend is sending the IPP options only if it's a PDF or an Image, but not when it's a Postscript file. In this case, even the Samsung printer evaluates the `NumCopies` from the Postscript file correctly and prints the correct amount of copies, because no IPP attributes are set.

| Printer | Copy parameter of IPP backend | Postscript NumCopies |Actually printed copies|
|---|---|---|---|
|  Samsung M5370LX  | not sent | 2 | 2 |
|  Samsung M4250LX | not sent | 2 | 2 |
| HP Laserjet M553  | not sent | 2 | 2 |

From my point of view, it's not a bug related to Papercut, but looks more like inconsistencies in the IPP-Backend. As a user, I would expect that the behavior of stdin and the file parameter is completely the same.

Is there a good reason why the behavior is different? Should PaperCut use stdin instead of the file parameter?

**System Information:**
 - OS and its version: Ubuntu 22.04
 - Application: Chrome Printing on Windows 10 using Build-In IPP
 - CUPS version: 2.4.2 compiled from source using debug flags

**Comment by michaelrsweet** [comment]
Short answer: the IPP backend (just like all other backends) is not intended to be used outside of CUPS. If you *do* use it this way, you need to reproduce how cupsd does things, which means passing your processed print data via stdin (either a pipe or just redirect from the file) to get the desired behavior.

Not going to change this behavior because it will break shared printing (where the options need to be sent to the remote CUPS server for processing).

**Comment by SoerenBusse** [comment]
Thank you very much for the clarification.

**Comment by SoerenBusse** [comment]
In case somebody is stumpling across this issue later:

Papercut implemented a config entry in version 22.0.11 to preverse the behavior when calling the original CUPS Backend. Add the following to the print-provider.conf:

```
PreverseCupsBackendCallBehavior=true
```
(Yes it's behavior, although the documentation states behaviour)
